// tspl-generator/dist/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BarcodeType = exports.Rotation = exports.Font = exports.MeasurementSystem = void 0;
/**
 * Measurement system options
 */
var MeasurementSystem;
(function (MeasurementSystem) {
    MeasurementSystem["ENGLISH"] = "ENGLISH";
    MeasurementSystem["METRIC"] = "METRIC";
})(MeasurementSystem || (exports.MeasurementSystem = MeasurementSystem = {}));
/**
 * Font options for TSPL
 */
var Font;
(function (Font) {
    Font["FONT_1"] = "1";
    Font["FONT_2"] = "2";
    Font["FONT_3"] = "3";
    // Add more fonts as needed
})(Font || (exports.Font = Font = {}));
/**
 * Rotation options
 */
var Rotation;
(function (Rotation) {
    Rotation[Rotation["NO_ROTATION"] = 0] = "NO_ROTATION";
    Rotation[Rotation["ROTATE_90"] = 90] = "ROTATE_90";
    Rotation[Rotation["ROTATE_180"] = 180] = "ROTATE_180";
    Rotation[Rotation["ROTATE_270"] = 270] = "ROTATE_270";
})(Rotation || (exports.Rotation = Rotation = {}));
/**
 * Barcode types supported by TSPL
 */
var BarcodeType;
(function (BarcodeType) {
    BarcodeType["CODE39"] = "39";
    BarcodeType["CODE128"] = "128";
    BarcodeType["EAN8"] = "8";
    BarcodeType["EAN13"] = "13";
    BarcodeType["QRCODE"] = "QRCODE";
    // Add more as needed
})(BarcodeType || (exports.BarcodeType = BarcodeType = {}));
