// tspl-generator/dist/printer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TSPLPrinter = void 0;
const commands = __importStar(require("./commands"));
const types_1 = require("./types");
/**
 * TSPL Printer class for generating TSPL commands
 */
class TSPLPrinter {
    /**
     * Create a new TSPL printer instance
     * @param measurementSystem - The measurement system to use
     */
    constructor(measurementSystem = types_1.MeasurementSystem.ENGLISH) {
        this.buffer = "";
        this.measurementSystem = measurementSystem;
    }
    /**
     * Initialize the printer with basic settings
     * @param config - Label configuration
     * @returns The printer instance for method chaining
     */
    initialize(config) {
        const { width, height, speed = 3, density = 8, gap = 3, gapOffset = 0, measurementSystem = this.measurementSystem, } = config;
        this.buffer = "";
        this.buffer += commands.size(width, height, measurementSystem);
        this.buffer += commands.speed(speed);
        this.buffer += commands.density(density);
        this.buffer += commands.gap(gap, gapOffset, measurementSystem);
        this.buffer += commands.cls();
        return this;
    }
    /**
     * Set the gap between labels
     * @param gap - Gap distance
     * @param offset - Offset distance
     * @returns The printer instance for method chaining
     */
    setGap(gap, offset = 0) {
        this.buffer += commands.gap(gap, offset, this.measurementSystem);
        return this;
    }
    /**
     * Add text to the label
     * @param options - Text options
     * @returns The printer instance for method chaining
     */
    addText(options) {
        const { x, y, font, rotation = 0, xMultiplier = 1, yMultiplier = 1, text, } = options;
        this.buffer += commands.text(x, y, font, rotation, xMultiplier, yMultiplier, text);
        return this;
    }
    /**
     * Add a text block to the label
     * @param x - X coordinate
     * @param y - Y coordinate
     * @param width - Width of the block
     * @param height - Height of the block
     * @param font - Font name or number
     * @param text - Text content
     * @param rotation - Rotation angle
     * @param xMultiplier - Horizontal multiplier
     * @param yMultiplier - Vertical multiplier
     * @param lineSpacing - Line spacing
     * @param alignment - Text alignment
     * @returns The printer instance for method chaining
     */
    addTextBlock(x, y, width, height, font, text, rotation = types_1.Rotation.NO_ROTATION, xMultiplier = 1, yMultiplier = 1, lineSpacing = 0, alignment = "L") {
        this.buffer += commands.block(x, y, width, height, font, rotation, xMultiplier, yMultiplier, lineSpacing, alignment, text);
        return this;
    }
    /**
     * Add a barcode to the label
     * @param x - X coordinate
     * @param y - Y coordinate
     * @param barcodeType - Type of barcode
     * @param height - Height of barcode
     * @param content - Barcode content
     * @param readable - Human readable text
     * @param rotation - Rotation angle
     * @param narrow - Narrow bar width
     * @param wide - Wide bar width
     * @returns The printer instance for method chaining
     */
    addBarcode(x, y, barcodeType, height, content, readable = 1, rotation = types_1.Rotation.NO_ROTATION, narrow = 2, wide = 4) {
        this.buffer += commands.barcode(x, y, barcodeType, height, readable, rotation, narrow, wide, content);
        return this;
    }
    /**
     * Add a QR code to the label
     * @param x - X coordinate
     * @param y - Y coordinate
     * @param content - QR code content
     * @param eccLevel - Error correction level
     * @param cellWidth - Cell width
     * @param mode - Mode
     * @param rotation - Rotation angle
     * @returns The printer instance for method chaining
     */
    addQRCode(x, y, content, eccLevel = "M", cellWidth = 6, mode = "A", rotation = types_1.Rotation.NO_ROTATION) {
        this.buffer += commands.qrcode(x, y, eccLevel, cellWidth, mode, rotation, content);
        return this;
    }
    /**
     * Add a box to the label
     * @param x - X coordinate of top-left corner
     * @param y - Y coordinate of top-left corner
     * @param xEnd - X coordinate of bottom-right corner
     * @param yEnd - Y coordinate of bottom-right corner
     * @param thickness - Line thickness
     * @returns The printer instance for method chaining
     */
    addBox(x, y, xEnd, yEnd, thickness = 1) {
        this.buffer += commands.box(x, y, xEnd, yEnd, thickness);
        return this;
    }
    /**
     * Add a line to the label
     * @param x - X coordinate of start point
     * @param y - Y coordinate of start point
     * @param xEnd - X coordinate of end point
     * @param yEnd - Y coordinate of end point
     * @param thickness - Line thickness
     * @returns The printer instance for method chaining
     */
    addLine(x, y, xEnd, yEnd, thickness = 1) {
        this.buffer += commands.line(x, y, xEnd, yEnd, thickness);
        return this;
    }
    /**
     * Add a circle to the label
     * @param x - X coordinate of center
     * @param y - Y coordinate of center
     * @param diameter - Diameter of circle
     * @param thickness - Line thickness
     * @returns The printer instance for method chaining
     */
    addCircle(x, y, diameter, thickness = 1) {
        this.buffer += commands.circle(x, y, diameter, thickness);
        return this;
    }
    /**
     * Add an ellipse to the label
     * @param x - X coordinate of center
     * @param y - Y coordinate of center
     * @param width - Width of ellipse
     * @param height - Height of ellipse
     * @param thickness - Line thickness
     * @returns The printer instance for method chaining
     */
    addEllipse(x, y, width, height, thickness = 1) {
        this.buffer += commands.ellipse(x, y, width, height, thickness);
        return this;
    }
    /**
     * Add a bitmap image to the label
     * @param x - X coordinate
     * @param y - Y coordinate
     * @param width - Width of bitmap in bytes
     * @param height - Height of bitmap in dots
     * @param mode - Mode (0: normal, 1: XOR)
     * @param bitmap - Bitmap data in hexadecimal format
     * @returns The printer instance for method chaining
     */
    addBitmap(x, y, width, height, bitmap, mode = 0) {
        this.buffer += commands.bitmap(x, y, width, height, mode, bitmap);
        return this;
    }
    /**
     * Add a BMP image to the label
     * @param x - X coordinate
     * @param y - Y coordinate
     * @param filename - BMP filename
     * @returns The printer instance for method chaining
     */
    addBMP(x, y, filename) {
        this.buffer += commands.putbmp(x, y, filename);
        return this;
    }
    /**
     * Print the label
     * @param copies - Number of copies to print
     * @returns The printer instance for method chaining
     */
    print(copies = 1) {
        this.buffer += commands.print(copies);
        return this;
    }
    /**
     * Clear the image buffer
     * @returns The printer instance for method chaining
     */
    clear() {
        this.buffer += commands.cls();
        return this;
    }
    /**
     * Get the generated TSPL code
     * @returns The complete TSPL code
     */
    getBuffer() {
        return this.buffer;
    }
    /**
     * Reset the buffer
     * @returns The printer instance for method chaining
     */
    reset() {
        this.buffer = "";
        return this;
    }
}
exports.TSPLPrinter = TSPLPrinter;
